sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3071,15 +3071,31 @@ class Analyzer(
       case p => p transformExpressions {
         case u @ UpCast(child, _, _) if !child.resolved => u
 
-        case UpCast(child, dt: AtomicType, _)
+        case u @ UpCast(child, _, _)
             if SQLConf.get.getConf(SQLConf.LEGACY_LOOSE_UPCAST) &&
+              u.dataType.isInstanceOf[AtomicType] &&
               child.dataType == StringType =>
-          Cast(child, dt.asNullable)
-
-        case UpCast(child, dataType, walkedTypePath) if !Cast.canUpCast(child.dataType, dataType) =>
-          fail(child, dataType, walkedTypePath)
-
-        case UpCast(child, dataType, _) => Cast(child, dataType.asNullable)
+          Cast(child, u.dataType.asNullable)
+
+        case u @ UpCast(child, _, walkedTypePath)
+          if child.dataType.isInstanceOf[DecimalType]
+            && u.dataType.isInstanceOf[DecimalType]
+            && walkedTypePath.nonEmpty =>
```

Comment:
why do we need to check `walkedTypePath`?

Reply:
It's more like a sanity check to prevent the case of using `UpCast` outside of `ExpressionEncoder` in the future.

Reply:
then let's add an assert, to make sure other code paths don't use `DecimalType` object as target type.

Reply:
+1